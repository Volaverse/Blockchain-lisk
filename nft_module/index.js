const { BaseModule } = require("lisk-sdk");
const { getAllNFTTokensAsJSON } = require("./nft");

const CreateNFTAsset = require("./transactions/create_nft_asset");
const PurchaseNFTAsset = require("./transactions/purchase_nft_asset");
const TransferNFTAsset = require("./transactions/transfer_nft_asset");
const TransferTestNFTAsset = require("./transactions/transfer_test_nft_asset");
const SellNFTAsset = require("./transactions/sell_nft_asset");
const EditNFTNameAsset = require("./transactions/edit_title_nft_asset")
const EditNFTDescAsset = require("./transactions/edit_description_nft_asset")
// Extend base module to implement your custom module
class NFTModule extends BaseModule {
  name = "nft";
  id = 1024;
  accountSchema = {
    type: "object",
    required: ["ownNFTs"],
    properties: {
      ownNFTs: {
        type: "array",
        fieldNumber: 1,
        items: {
          dataType: "bytes",
        },
      },
    },
    default: {
      ownNFTs: [],
    },
  };
  transactionAssets = [new CreateNFTAsset(), new PurchaseNFTAsset(), new TransferNFTAsset(),new TransferTestNFTAsset(),new SellNFTAsset(),new EditNFTNameAsset(),new EditNFTDescAsset()];
  actions = {
    // get all the registered NFT tokens from blockchain
    getAllNFTTokens: async () => getAllNFTTokensAsJSON(this._dataAccess),
  };
}

module.exports = { NFTModule };
