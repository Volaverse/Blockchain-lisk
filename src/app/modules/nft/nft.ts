/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */

import { codec, cryptography } from 'lisk-sdk';

export interface nftInterface {
	id: {
		dataType: 'bytes';
	};
	value: {
		dataType: 'uint64';
	};
	ownerAddress: {
		dataType: 'bytes';
	};
	minPurchaseMargin: {
		dataType: 'uint32';
	};
	name: {
		dataType: 'string';
	};
	description: {
		dataType: 'string';
	};
	category: {
		dataType: 'uint32';
	};
}
interface RegisteredNFT {
	registeredNFTTokens: nftInterface[];
}

export const registeredNFTTokensSchema = {
	$id: 'lisk/nft/registeredTokens',
	type: 'object',
	required: ['registeredNFTTokens'],
	properties: {
		registeredNFTTokens: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				required: ['id', 'value', 'ownerAddress', 'minPurchaseMargin', 'name', 'category'],
				properties: {
					id: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					value: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
					ownerAddress: {
						dataType: 'bytes',
						fieldNumber: 3,
					},
					minPurchaseMargin: {
						dataType: 'uint32',
						fieldNumber: 4,
					},
					name: {
						dataType: 'string',
						fieldNumber: 5,
					},
					description: {
						dataType: 'string',
						fieldNumber: 6,
					},
					category: {
						dataType: 'uint32',
						fieldNumber: 7,
					},
				},
			},
		},
	},
};

export const CHAIN_STATE_NFT_TOKENS = 'nft:registeredNFTTokens';

// returns nft object with id
export const createNFTToken = ({
	name,
	description,
	ownerAddress,
	nonce,
	value,
	minPurchaseMargin,
	category,
}) => {
	const nonceBuffer = Buffer.alloc(8);
	nonceBuffer.writeBigInt64LE(nonce);
	const seed = Buffer.concat([ownerAddress, nonceBuffer]);
	const id = cryptography.hash(seed);

	return {
		id,
		minPurchaseMargin,
		name,
		description,
		ownerAddress,
		value,
		category,
	};
};
// get list of all nft tokens
export const getAllNFTTokens = async stateStore => {
	const registeredTokensBuffer = await stateStore.chain.get(CHAIN_STATE_NFT_TOKENS);
	if (!registeredTokensBuffer) {
		return [];
	}

	const registeredTokens: any = codec.decode(registeredNFTTokensSchema, registeredTokensBuffer);

	return registeredTokens.registeredNFTTokens;
};
// returns nft tokens as json
export const getAllNFTTokensAsJSON = async dataAccess => {
	const registeredTokensBuffer = await dataAccess.getChainState(CHAIN_STATE_NFT_TOKENS);

	if (!registeredTokensBuffer) {
		return [];
	}

	const registeredTokens = codec.decode<RegisteredNFT>(
		registeredNFTTokensSchema,
		registeredTokensBuffer,
	);

	return codec.toJSON<RegisteredNFT>(registeredNFTTokensSchema, registeredTokens)
		.registeredNFTTokens;
};
// set nft tokens to state store
export const setAllNFTTokens = async (stateStore, NFTTokens) => {
	const registeredTokens = {
		registeredNFTTokens: NFTTokens.sort((a, b) => a.id.compare(b.id)),
	};

	await stateStore.chain.set(
		CHAIN_STATE_NFT_TOKENS,
		codec.encode(registeredNFTTokensSchema, registeredTokens),
	);
};

module.exports = {
	registeredNFTTokensSchema,
	CHAIN_STATE_NFT_TOKENS,
	getAllNFTTokens,
	setAllNFTTokens,
	getAllNFTTokensAsJSON,
	createNFTToken,
};
