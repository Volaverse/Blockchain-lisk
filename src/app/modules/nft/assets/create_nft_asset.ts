/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

import { BaseAsset } from 'lisk-sdk';
import { getAllNFTTokens, setAllNFTTokens, createNFTToken } from '../nft';

export class CreateNftAsset extends BaseAsset {
	public name = 'createNft';
	public id = 0;

	// define schema for asset
	public schema = {
		$id: 'nft/createNft-asset',
		title: 'CreateNftAsset transaction asset for nft module',
		type: 'object',
		required: ['minPurchaseMargin', 'initValue', 'name', 'description', 'category'],
		properties: {
			minPurchaseMargin: {
				dataType: 'uint32',
				fieldNumber: 1,
			},
			initValue: {
				dataType: 'uint64',
				fieldNumber: 2,
			},
			name: {
				dataType: 'string',
				fieldNumber: 3,
			},
			description: {
				dataType: 'string',
				fieldNumber: 4,
			},
			category: {
				dataType: 'uint32',
				fieldNumber: 5,
			},
		},
	};

	public validate({ asset }): void {
		// validate your asset
		if (asset.initValue <= 0) {
			throw new Error('NFT init value is too low.');
		} else if (asset.minPurchaseMargin < 0 || asset.minPurchaseMargin > 100) {
			throw new Error('The NFT minimum purchase value needs to be between 0 and 100.');
		}
	}

	public async apply({ asset, transaction, stateStore, reducerHandler }): Promise<void> {
		const allTokens = await getAllNFTTokens(stateStore);
		const { senderAddress } = transaction;
		const senderAccount = await stateStore.account.get(senderAddress);
		let dup = [];
		// Checking for unique nft name
		if (allTokens.length > 0) {
			dup = allTokens.filter(item => item.name === asset.name);
		}

		if (dup.length > 0) {
			throw new Error('Name of nft should be unique');
		}
		// category 3 is for usernameNFT which anyone can create
		// Other categories NFT can only be created by a particular address
		if (
			senderAddress.toString('hex') !== '16c70194f16fa137d96168823f695d2ddb232554' &&
			!(asset.category === 3 || asset.category === '3')
		) {
			throw new Error(`NFT cannot be created from this account ${senderAddress.toString('hex')}`);
		}
		// create nft
		const nftToken = createNFTToken({
			name: asset.name,
			description: asset.description,
			ownerAddress: senderAddress,
			nonce: transaction.nonce,
			value: asset.initValue,
			minPurchaseMargin: asset.minPurchaseMargin,
			category: asset.category,
		});

		// 6.update sender account with unique nft id
		senderAccount.nft.ownNFTs.push(nftToken.id);
		await stateStore.account.set(senderAddress, senderAccount);

		// 7.debit tokens from sender account to create nft
		await reducerHandler.invoke('token:debit', {
			address: senderAddress,
			amount: asset.initValue,
		});

		// 8.save nfts

		allTokens.push(nftToken);
		await setAllNFTTokens(stateStore, allTokens);
	}
}
