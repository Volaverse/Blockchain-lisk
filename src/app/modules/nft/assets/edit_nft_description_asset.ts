/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { BaseAsset } from 'lisk-sdk';
import { getAllNFTTokens, setAllNFTTokens } from '../nft';

export class EditNftDescriptionAsset extends BaseAsset {
	public name = 'editNftDescription';
	public id = 4;

	// define schema for asset
	public schema = {
		$id: 'nft/editNftDescription-asset',
		title: 'EditNftDescriptionAsset transaction asset for nft module',
		type: 'object',
		required: ['nftId', 'description'],
		properties: {
			nftId: {
				dataType: 'bytes',
				fieldNumber: 1,
			},
			description: {
				dataType: 'string',
				fieldNumber: 2,
			},
		},
	};

	public validate({ asset }): void {
		if (asset.nftId === '' || !asset.nftId) {
			throw new Error('NFT id should have value');
		}
		if (asset.description === '' || !asset.description) {
			throw new Error('description cannot be empty');
		}
	}

	public async apply({ asset, transaction, stateStore }): Promise<void> {
		const nftTokens = await getAllNFTTokens(stateStore);
		const nftTokenIndex = nftTokens.findIndex(t => t.id.equals(asset.nftId));

		// verify if the nft exists
		if (nftTokenIndex < 0) {
			throw new Error('Token id not found');
		}

		const token = nftTokens[nftTokenIndex];
		const tokenOwnerAddress = token.ownerAddress;
		const { senderAddress } = transaction;
		// verify that the sender owns the nft

		if (!tokenOwnerAddress.equals(senderAddress)) {
			throw new Error('An NFT description can only be changed by the owner of the NFT.');
		}

		// setting the minimum sell value
		token.description = asset.description;
		nftTokens[nftTokenIndex] = token;
		await setAllNFTTokens(stateStore, nftTokens);
	}
}
