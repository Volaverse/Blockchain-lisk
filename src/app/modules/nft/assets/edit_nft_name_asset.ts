/* eslint-disable @typescript-eslint/restrict-plus-operands */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { BaseAsset } from "lisk-sdk";
import { getAllNFTTokens, setAllNFTTokens } from "../nft";

export class EditNftNameAsset extends BaseAsset {
  public name = "editNftName";
  public id = 3;

  // define schema for asset
  public schema = {
    $id: "nft/editNftName-asset",
    title: "EditNftNameAsset transaction asset for nft module",
    type: "object",
    required: ["nftId", "name"],
    properties: {
      nftId: {
        dataType: "bytes",
        fieldNumber: 1,
      },
      name: {
        dataType: "string",
        fieldNumber: 2,
      },
    },
  };

  public validate({ asset }): void {
    if (asset.nftId === "" || !asset.nftId) {
      throw new Error("NFT id should have value");
    }
    if (asset.name === "" || !asset.name) {
      throw new Error("description cannot be empty");
    }
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }): Promise<void> {
    const nftTokens = await getAllNFTTokens(stateStore);
    const nftTokenIndex = nftTokens.findIndex((t) => t.id.equals(asset.nftId));

    if (asset.name === "") {
      throw new Error("Name cannot be empty");
    }

    // verify if the nft exists
    if (nftTokenIndex < 0) {
      throw new Error("Token id not found");
    }
    // checking for unique nft name
    let dup = [];
    if (nftTokens.length > 0) {
      dup = nftTokens.filter((item) => item.name === asset.name);
    }

    if (dup.length > 0) {
      throw new Error("Name of nft should be unique");
    }
    const token = nftTokens[nftTokenIndex];
    if (token.category !== 3) {
      throw new Error("Name can only be changed of username category");
    }
    const tokenOwnerAddress = token.ownerAddress;
    const { senderAddress } = transaction;
    // verify that the sender owns the nft

    if (!tokenOwnerAddress.equals(senderAddress)) {
      throw new Error("An NFT can only be renamed by the owner of the NFT.");
    }

    // setting the minimum sell value
    token.name = asset.name;
    nftTokens[nftTokenIndex] = token;
    await setAllNFTTokens(stateStore, nftTokens);
  }
}
