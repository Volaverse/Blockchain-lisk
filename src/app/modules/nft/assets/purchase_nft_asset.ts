/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable @typescript-eslint/restrict-plus-operands */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { BaseAsset } from 'lisk-sdk';

import { getAllNFTTokens, setAllNFTTokens } from '../nft';

export class PurchaseNftAsset extends BaseAsset {
	public name = 'purchaseNft';
	public id = 1;

	// define schema for asset
	public schema = {
		$id: 'nft/purchaseNft-asset',
		title: 'PurchaseNftAsset transaction asset for nft module',
		type: 'object',
		required: ['nftId', 'purchaseValue'],
		properties: {
			nftId: {
				dataType: 'bytes',
				fieldNumber: 1,
			},
			purchaseValue: {
				dataType: 'uint64',
				fieldNumber: 2,
			},
			name: {
				dataType: 'string',
				fieldNumber: 3,
			},
		},
	};

	public validate({ asset }): void {
		// validate your asset
		if (asset.nftId === '' || !asset.nftId) {
			throw new Error('NFT id should have value');
		}
		if (asset.purchaseValue === '' || !asset.purchaseValue) {
			throw new Error('purchaseValue should have value');
		}
		if (asset.purchaseValue <= 0) {
			throw new Error('The NFT purchase value needs to be more than 0');
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore, reducerHandler }): Promise<void> {
		const nftTokens = await getAllNFTTokens(stateStore);
		const nftTokenIndex = nftTokens.findIndex(t => t.id.equals(asset.nftId));

		// verify if purchasing nft exists
		if (nftTokenIndex < 0) {
			throw new Error('Token id not found');
		}
		const token = nftTokens[nftTokenIndex];
		const tokenOwner = await stateStore.account.get(token.ownerAddress);
		const tokenOwnerAddress = tokenOwner.address;

		// 5.verify if minimum nft purchasing condition met
		if (token && token.minPurchaseMargin === 0) {
			throw new Error('This NFT token can not be purchased');
		}

		const tokenCurrentValue = token.value;
		const tokenMinPurchaseValue =
			tokenCurrentValue + (tokenCurrentValue * BigInt(token.minPurchaseMargin)) / BigInt(100);
		const { purchaseValue } = asset;

		if (tokenMinPurchaseValue > purchaseValue) {
			throw new Error(
				`Token can not be purchased. Purchase value is too low. Minimum value: ${tokenMinPurchaseValue}`,
			);
		}

		const purchaserAddress = transaction.senderAddress;
		const purchaserAccount = await stateStore.account.get(purchaserAddress);

		// remove nft from owner account
		const ownerTokenIndex = tokenOwner.nft.ownNFTs.findIndex(a => a.equals(token.id));
		tokenOwner.nft.ownNFTs.splice(ownerTokenIndex, 1);
		await stateStore.account.set(tokenOwnerAddress, tokenOwner);

		// add nft to purchaser account
		purchaserAccount.nft.ownNFTs.push(token.id);
		await stateStore.account.set(purchaserAddress, purchaserAccount);

		token.ownerAddress = purchaserAddress;
		token.value = purchaseValue;
		token.minPurchaseMargin = 0;
		nftTokens[nftTokenIndex] = token;
		await setAllNFTTokens(stateStore, nftTokens);

		// debit LSK tokens from purchaser account
		await reducerHandler.invoke('token:debit', {
			address: purchaserAddress,
			amount: purchaseValue,
		});

		// credit LSK tokens to purchaser account
		await reducerHandler.invoke('token:credit', {
			address: tokenOwnerAddress,
			amount: purchaseValue,
		});
	}
}
