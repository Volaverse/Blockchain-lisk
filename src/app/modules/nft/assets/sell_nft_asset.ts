/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable eqeqeq */
import { BaseAsset } from 'lisk-sdk';
import { getAllNFTTokens, setAllNFTTokens } from '../nft';

export class SellNftAsset extends BaseAsset {
	public name = 'sellNft';
	public id = 2;

	// define schema for asset
	public schema = {
		$id: 'nft/sellNft-asset',
		title: 'SellNftAsset transaction asset for nft module',
		type: 'object',
		required: ['nftId', 'minPurchaseMargin'],
		properties: {
			nftId: {
				dataType: 'bytes',
				fieldNumber: 1,
			},
			minPurchaseMargin: {
				dataType: 'uint32',
				fieldNumber: 2,
			},
			name: {
				dataType: 'string',
				fieldNumber: 3,
			},
		},
	};

	public validate({ asset }): void {
		if (asset.nftId == '' || !asset.nftId) {
			throw new Error('NFT id should have value');
		}
		if (asset.minPurchaseMargin == '' || !asset.minPurchaseMargin) {
			throw new Error('NFT id should have value');
		}
		if (asset.minPurchaseMargin < 0) {
			throw new Error('Minimum Purchase Margin value should be more than 0');
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore }): Promise<void> {
		const nftTokens = await getAllNFTTokens(stateStore);
		const nftTokenIndex = nftTokens.findIndex(t => t.id.equals(asset.nftId));

		// verify if the nft exists
		if (nftTokenIndex < 0) {
			throw new Error('Token id not found');
		}
		if (asset.minPurchaseMargin < 0 || asset.minPurchaseMargin > 100) {
			throw new Error('The NFT minimum purchase value needs to be between 0 and 100.');
		}
		const token = nftTokens[nftTokenIndex];
		const tokenOwnerAddress = token.ownerAddress;
		const { senderAddress } = transaction;
		// verify that the sender owns the nft

		if (!tokenOwnerAddress.equals(senderAddress)) {
			throw new Error('An NFT can only be sold by the owner of the NFT.');
		}

		// setting the minimum sell value

		token.minPurchaseMargin = asset.minPurchaseMargin;
		nftTokens[nftTokenIndex] = token;
		await setAllNFTTokens(stateStore, nftTokens);
	}
}
