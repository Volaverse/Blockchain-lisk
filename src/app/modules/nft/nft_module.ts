/* eslint-disable class-methods-use-this */

import { BaseModule } from 'lisk-sdk';
import { CreateNftAsset } from './assets/create_nft_asset';
import { EditNftDescriptionAsset } from './assets/edit_nft_description_asset';
import { EditNftNameAsset } from './assets/edit_nft_name_asset';
import { PurchaseNftAsset } from './assets/purchase_nft_asset';
import { SellNftAsset } from './assets/sell_nft_asset';
import { getAllNFTTokensAsJSON } from './nft';

export class NftModule extends BaseModule {
	public actions = {
		getAllNFTTokens: async () => getAllNFTTokensAsJSON(this._dataAccess),
	};
	public reducers = {};
	public name = 'nft';
	public transactionAssets = [
		new CreateNftAsset(),
		new PurchaseNftAsset(),
		new SellNftAsset(),
		new EditNftNameAsset(),
		new EditNftDescriptionAsset(),
	];
	public events = [];
	public id = 1024;
	public accountSchema = {
		type: 'object',
		required: ['ownNFTs'],
		properties: {
			ownNFTs: {
				type: 'array',
				fieldNumber: 1,
				items: {
					dataType: 'bytes',
				},
			},
		},
		default: {
			ownNFTs: [],
		},
	};
}
