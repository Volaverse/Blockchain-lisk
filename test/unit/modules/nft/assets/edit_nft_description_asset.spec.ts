import { ReducerHandler, StateStore, codec, testing } from "lisk-sdk";
import { EditNftDescriptionAsset } from "../../../../../src/app/modules/nft/assets/edit_nft_description_asset";
import { NftModule } from "../../../../../src/app/modules/nft/nft_module";
import {
  createNFTToken,
  getAllNFTTokens,
  registeredNFTTokensSchema,
} from "../../../../../src/app/modules/nft/nft";

describe("EditNftDescriptionAsset", () => {
  let transactionAsset: EditNftDescriptionAsset;

  beforeEach(() => {
    transactionAsset = new EditNftDescriptionAsset();
  });

  describe("constructor", () => {
    it("should have valid id", () => {
      expect(transactionAsset.id).toEqual(4);
    });

    it("should have valid name", () => {
      expect(transactionAsset.name).toEqual("editNftDescription");
    });

    it("should have valid schema", () => {
      expect(transactionAsset.schema).toMatchSnapshot();
    });
  });

  describe("validate", () => {
    describe("schema validation", () => {
      it("should throw errors for missing NFT id", () => {
        const context = testing.createValidateAssetContext({
          asset: {
            description: "description 1",
          },
          transaction: { senderAddress: Buffer.alloc(0) } as any,
        });

        expect(() => transactionAsset.validate(context)).toThrow();
      });
      it("should throw errors for missing  description", () => {
        const context = testing.createValidateAssetContext({
          asset: {
            nftId: "2dfr1",
          },
          transaction: { senderAddress: Buffer.alloc(0) } as any,
        });

        expect(() => transactionAsset.validate(context)).toThrow();
      });
      it("should pass on validate asset", () => {
        const context = testing.createValidateAssetContext({
          asset: {
            nftId: "2dfr1",
            description: "description 1",
          },
          transaction: { senderAddress: Buffer.alloc(0) } as any,
        });

        expect(() => transactionAsset.validate(context)).not.toThrow();
      });
    });
  });

  describe("apply", () => {
    let stateStore: StateStore;
    let reducerHandler: ReducerHandler;
    let account: any;
    let context;
    let invalidContext;
    let nft;
    let nftObj;

    beforeEach(async () => {
      account = testing.fixtures.createDefaultAccount([NftModule]);
      nft = createNFTToken({
        value: BigInt("100"),
        ownerAddress: account.address,
        minPurchaseMargin: 1,
        name: "nft 1",
        description: "description",
        category: 3,
        nonce: BigInt(0),
        imageUrl:"imageUrl",
        threeDUrl:"threeDUrl"
      });
      nftObj = {
        registeredNFTTokens: [nft],
      };

      stateStore = new testing.mocks.StateStoreMock({
        accounts: [account],
      });

      await stateStore.chain.set(
        "nft:registeredNFTTokens",
        codec.encode(registeredNFTTokensSchema, nftObj)
      );

      reducerHandler = testing.mocks.reducerHandlerMock;
      context = testing.createApplyAssetContext({
        stateStore,
        reducerHandler,
        asset: {
          nftId: nft.id,
          description: "updated description",
        },
        transaction: {
          senderAddress: account.address,
          nonce: BigInt(0),
        } as any,
      });
      invalidContext = testing.createApplyAssetContext({
        stateStore,
        reducerHandler,
        asset: {
          nftId: nft.id,
          description: "updated description",
        },
        transaction: {
          senderAddress: Buffer.alloc(0),
          nonce: BigInt(0),
        } as any,
      });

      jest.spyOn(stateStore.chain, "get");
      jest.spyOn(stateStore.chain, "set");
      jest.spyOn(reducerHandler, "invoke");
    });

    describe("valid cases", () => {
      it("should edit name of the nft", async () => {
        await transactionAsset.apply(context);
        const tokens = await getAllNFTTokens(stateStore);

        expect(tokens[0].description).toEqual("updated description");
      });
    });

    describe("invalid cases", () => {
      it("it should throw error as only owner can change the description", async () => {
        await expect(transactionAsset.apply(invalidContext)).rejects.toThrow();
      });
    });
  });
});
