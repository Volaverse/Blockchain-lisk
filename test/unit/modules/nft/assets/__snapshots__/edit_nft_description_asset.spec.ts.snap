// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditNftDescriptionAsset constructor should have valid schema 1`] = `
Object {
  "$id": "nft/editNftDescription-asset",
  "properties": Object {
    "description": Object {
      "dataType": "string",
      "fieldNumber": 2,
    },
    "nftId": Object {
      "dataType": "bytes",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "nftId",
    "description",
  ],
  "title": "EditNftDescriptionAsset transaction asset for nft module",
  "type": "object",
}
`;
